python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,182 @@
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, client_kwargs: Dict[str, str] = EMPTY_DICT, **fs_kwargs):
+        client_properties = {f"client_{key}": val for key, val in client_kwargs.items()}
+        fs_properties = {f"fs_{key}": val for key, val in fs_kwargs.items()}
+        properties = {**client_properties, **fs_properties}
+        self._fs = None
+        super().__init__(properties=properties)
+
+    def new_input(self, location: str) -> FsspecInputFile:
+        """Get an FsspecInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecInputFile: An FsspecInputFile instance for the given location
+        """
+        uri = urlparse(location)
+        fs = self.properties.get("fs") or self._resolve_fs_type(scheme=uri.scheme)
+        return FsspecInputFile(location=location, fs=fs)
+
+    def new_output(self, location: str) -> FsspecOutputFile:
+        """Get an FsspecOutputFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecOutputFile: An FsspecOutputFile instance for the given location
+        """
+        uri = urlparse(location)
+        fs = self.properties.get("fs") or self._resolve_fs_type(scheme=uri.scheme)
+        return FsspecOutputFile(location=location, fs=fs)
+
+    def delete(self, location: Union[str, InputFile, OutputFile]) -> None:
+        """Delete the file at the given location
+
+        Args:
+            location(str, InputFile, OutputFile): The URI to the file--if an InputFile instance or an
+            OutputFile instance is provided, the location attribute for that instance is used as the location
+            to delete
+        """
+        if isinstance(location, (InputFile, OutputFile)):
+            str_location = location.location  # Use InputFile or OutputFile location
+        else:
+            str_location = location
+
+        uri = urlparse(str_location)
+        fs = self._fs or self._resolve_fs_type(scheme=uri.scheme)
+        fs.rm(str_location)
+
+    def _resolve_fs_type(self, scheme: str):
+        """Initialize the proper filesystem for the location's scheme"""
+        if scheme in ("s3", "s3n", "s3a"):
+            from s3fs import S3FileSystem
+
+            self._fs = S3FileSystem(client_kwargs=self._client_properties() or None, **self._fs_properties())
+            return self._fs
+        else:
+            raise ValueError(f"Cannot determine fsspec implementation to use with scheme: {scheme}")
+
+    def _fs_properties(self):
+        """Get fs properties from the file-io property map"""
+        return {k[3:]: v for k, v in self.properties.items() if k.startswith("fs_")}
```

Comment:
If we want to do it like this, then we should do something like:
```suggestion
FS_ = 'fs_'

...

        return {k[len(FS_):]: v for k, v in self.properties.items() if k.startswith(FS_)}
```
But I think we can make this nicer. In the configuration, we have a recursive dictionairy.

This would allow us to set properties like:
```
catalog:
  production:
   uri: https://...
   credentials: token
   py-file-io: pyiceberg.io.fsspec.FsspecFileIO
   fs:
     access_key: abc
     secret_key: def
   client_kwargs:
     client_url: https://...
```

We could leverage that to avoid parsing all the properties. WDYT?

Reply:
I agree that we should use configuration, but I think we should limit it to `Properties` so that it is compatible with the properties used by Java. The config for FileIO would be:

```yaml
catalog:
  production:
    uri: https://...
    credential: token
    s3.endpoint: https://...
    s3.access-key-id: abc
    s3.secret-access-key: def
```

Reply:
So should we not allow passing in additional filesystem-specific kwargs? [Here's](https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem) a list of the s3fs kwargs which has stuff like `default_block_size` and `version_awarebool`.

Reply:
We can create a prefix for passing implementation-specific arguments.